### PR TITLE
Support non-standard timestamps (SEC, MS, & NS) in the C API

### PR DESCRIPTION
**Pull request by jraymakers**
New C API types:
- `duckdb_timestamp_s`
- `duckdb_timestamp_ms`
- `duckdb_timestamp_ns`
- `duckdb_timestamp_ns_struct` (to hold `nanos`)

New C API functions:
- Conversion:
  - `duckdb_from_timestamp_s`, `duckdb_to_timestamp_s`, `duckdb_is_finite_timestamp_s`
  - `duckdb_from_timestamp_ms`, `duckdb_to_timestamp_ms`, `duckdb_is_finite_timestamp_ms`
  - `duckdb_from_timestamp_ns`, `duckdb_to_timestamp_ns`, `duckdb_is_finite_timestamp_ns`
- Binding:
  - `duckdb_bind_timestamp_s`
  - `duckdb_bind_timestamp_ms`
  - `duckdb_bind_timestamp_ns`
- Value creation/access:
  - `duckdb_create_timestamp_tz`, `duckdb_get_timestamp_tz`
  - `duckdb_create_timestamp_s`, `duckdb_get_timestamp_s`
  - `duckdb_create_timestamp_ms`, `duckdb_get_timestamp_ms`
  - `duckdb_create_timestamp_ns`, `duckdb_get_timestamp_ns`
- Appending:
  - `duckdb_append_timestamp_s`
  - `duckdb_append_timestamp_ms`
  - `duckdb_append_timestamp_ns`

New C API unit tests for all of the above, plus some new tests for appending TS SEC, MS, & NS values with the C++ Appender.

Note that, while I added new C API types for the non-standard timestamps, I didn't change how the C++ layer deals with these types. The C++ layer mostly uses a single type, `timestamp_t`, for these variants. This type has a `value` field that may hold seconds, milliseconds, microseconds, or nanoseconds. There are some more specific subtypes, `timestamp_tz_t`, `timestamp_ns_t`, `timestamp_ms_t`, and `timestamp_sec_t`, but these contain no additional code and are only used in a couple places. I experimented with changing the C++ layer to use these more specific types more, but it seems that would require touching much more code, and I ran into errors I didn't know how to address. Hence this PR focuses on exposing this functionality through the C API, without changing how the C++ layer functions.

Also:
- `duckdb_bind_timestamp_tz` already exists.
- I don't believe `duckdb_append_timestamp_tz` is needed, because `duckdb_append_timestamp` can be used for TIMESTAMP_TZ types. This is because the destination type is already known when appending, and TIMESTAMP and TIMESTAMP_TZ are both represented by `duckdb_timestamp` in the C API, so a theoretical `duckdb_append_timestamp_tz` would have the same signature and behavior as `duckdb_append_timestamp`.

### CODE REVIEW

**Comment by jraymakers** [comment]
I discovered this related [PR](https://github.com/duckdb/duckdb/pull/14758) as I was posting this one. There's some overlap, but this one take a slightly different approach - in introducing new C API types - and also covers more functionality.

**Comment by taniabogatsch** [comment]
Hi, @jraymakers - I think this PR supersedes #14758. I'll remove the timestamp-related changes from #14758, and then we can focus on your more thorough work here. :)

Look at this PR implementing the missing functionality in C++: https://github.com/duckdb/duckdb/pull/14818. There's still some work to be done on that one (reviews/feedback is welcome). Then, we should merge this PR **after** improving the C++ layer via that other PR.

File: src/main/capi/duckdb_value-c.cpp
Comment (taniabogatsch) [inline src/main/capi/duckdb_value-c.cpp:159]: All the new getters in this file need a `nullptr` check. And the respective test case.
```c
if (!val) {
	return the_function_return_type {0};
}
```
Later, in the `test/api/capi/test_capi_complex_types.cpp` tests.
```c
REQUIRE(duckdb_the_function_return_type(nullptr).field == 0);
```

File: test/api/capi/test_capi_appender.cpp
Comment (taniabogatsch) [inline test/api/capi/test_capi_appender.cpp:722]: `auto ts_s_out = result->Fetch<duckdb_timestamp_s>(0, 0);` should do the trick here and in the other tests.

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
I've also left a few comments. :) It's almost ready; the C++ changes will need to be updated in some places.

File: src/main/capi/datetime-c.cpp
Comment (jraymakers) [reply src/main/capi/datetime-c.cpp:198]: Hm, a lot of the conversion functions user here throw, including in existing functions (e.g. `Timestamp::Convert` used by `duckdb_from_timestamp`, `Timestamp::FromDatetime` used by `duckdb_to_timestamp`), so this affects more than just this function.

Many other functions in the C API return a `duckdb_state` to indicate success or error, and have a separate function to get the error string on the relevant object. But there isn't a relevant object in this case.

Looks like the `out_error` pattern has one precedent: `duckdb_open_ext`. I could follow that for the new functions I'm adding.

But fixing all such problems here would require making breaking changes to existing functions, or more likely adding new ones and deprecating the old ones. That's out of scope for this change.

File: test/api/capi/test_capi_appender.cpp
Comment (jraymakers) [reply test/api/capi/test_capi_appender.cpp:722]: Unfortunately it doesn't. This gave me quite a headache, actually.

I started out using that pattern, following how it's done elsewhere, but I found the resulting values were off. For example, for TIMESTAMP_S, I'd get back a value in microseconds (e.g. 1,000,000 times the value I expected).

I tracked this down to how the various overloads of CAPIResult::Fetch work. They all use either one of the (deprecated) `duckdb_value_xxx` functions, or the (deprecated) `duckdb_column_data` function. Since there aren't `duckdb_value_xxx` functions for the non-standard timestamp types, and I don't want to add new functions to a deprecated set, I tried to use `duckdb_column_data`. But this uses `DeprecatedMaterializeResult`, which, while materializing, does some casting, including converting the non-standard timestamp types to microsecond timestamps. Hence the unexpected values I was seeing.

I didn't want to invest in updating this deprecated path, and perhaps more importantly it would result in a behavior change visible through the C API (and that seems like a no-go). So I opted to fetch the data directly, rather than using the tester to fetch it.

File: src/main/capi/datetime-c.cpp
Comment (jraymakers) [reply src/main/capi/datetime-c.cpp:198]: It seems a more common pattern is to catch any errors returned by the C++ layer and just return DuckDBError or DuckDBSuccess, rather than try to transfer the error message. The error messages aren't especially useful for these conversion messages; they're all some variation on "overflow" or "out of range", or just "could not convert".

This also avoids the problem of memory management for `out_error`. The one place it is used (`duckdb_open_ext`) expects the caller to free it. That would be pretty awkward for the caller to have to worry about for these conversion functions.

Comment (jraymakers) [reply src/main/capi/datetime-c.cpp]: I've updated the NS conversion functions to have an out parameter for the result and return a duckdb_state to indicate success or failure. Let me know if this looks like a good pattern, and I'll update the functions for the other types accordingly.

File: src/include/duckdb.h
Comment (Mytherin) [inline src/include/duckdb.h:1366]: I think all of these conversion functions to and from the struct are a bit much - in general we're not really using the `duckdb_timestamp_struct` anymore, and it's trivial to convert between `duckdb_timestamp_s` and `duckdb_timestamp` manually anyway if you need to. Perhaps we can leave these out?

File: src/include/duckdb.h
Comment (Mytherin) [inline src/include/duckdb.h:3949]: Can we leave these out as well? The `duckdb_append_` methods are not intended to be complete. You can use `duckdb_append_data_chunk` to append data chunks containing any type within DuckDB (and we should add `duckdb_append_value` at some point to append a `duckdb_value`).

File: src/include/duckdb.h
Comment (Mytherin) [inline src/include/duckdb.h:1708]: Similarly we don't need these bind methods - we can use `duckdb_bind_value`

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR!

Looks good - but I would remove a bunch of methods from this PR. In general, these types are rather "fringe" types and not commonly used. The `duckdb_value` type can be used in various places if you want to bind/append this type, so all that is necessary is the methods to convert to/from a `duckdb_value` - and I think it's fine to have the `duckdb_finite_timestamp_XX` methods as well as they can help with reading the values from a data chunk.

File: src/main/capi/datetime-c.cpp
Comment (taniabogatsch) [reply src/main/capi/datetime-c.cpp]: > It seems a more common pattern is to catch any errors returned by the C++ layer and just return DuckDBError or DuckDBSuccess, rather than try to transfer the error message

Yes, in many places, we also extract the error message and store it in the wrapper structure (e.g., the appender wrapper). But I don't think we want to do that here.

> I've updated the NS conversion functions to have an out parameter for the result and return a duckdb_state to indicate success or failure. Let me know if this looks like a good pattern, and I'll update the functions for the other types accordingly - or try some other approach.

Yes, great. I think this is lightweight enough (no deallocations on error and such), and at the same time, there is a way to indicate a failure.

File: test/api/capi/test_capi_appender.cpp
Comment (taniabogatsch) [reply test/api/capi/test_capi_appender.cpp:721]: All right, I agree that we should not add more deprecated functions or change any deprecated paths. Let's keep it like this. :)

**Comment by jraymakers** [comment]
> Looks good - but I would remove a bunch of methods from this PR. In general, these types are rather "fringe" types and not commonly used. The `duckdb_value` type can be used in various places if you want to bind/append this type, so all that is necessary is the methods to convert to/from a `duckdb_value` - and I think it's fine to have the `duckdb_finite_timestamp_XX` methods as well as they can help with reading the values from a data chunk.

I agree we could take the approach you suggest for exposing these types: relying on `duckdb_value` instead of type-specific functions for each. But I wanted to explain my motivation for adding the functions, and what I see as the downsides of relying on `duckdb_value` for these cases.

My primary goal is to enable binding values of all types (to prepared statements), and also appending values of all types, through the C API. In addition, I have a more specific goal of supporting them in the new Node client ("Neo"), which is built on the C API.

For binding, supporting `duckdb_value` creation for each type is required, so that arbitrary nested types can be constructed. Since there's already a `duckdb_bind_value`, we could stop there. But it's awkward and inefficient to have to create a `duckdb_value` for each parameter, because doing so requires two extra calls: one to create the value, and one later to destroy it. It also requires a heap allocation that isn't necessary for the type-specific bind calls, which create Value instances on the stack. These extra calls and allocations become even more awkward through a higher-level language client such as the (new) Node API.

For appending, there isn't currently a `duckdb_append_value`. We could add one, as you suggest. But, in addition to the same awkward and inefficient additional calls and allocations as in the binding case, there's an additional awkwardness for the append case: creating a value requires the type to be specified again. For the bind case, it's necessary to specify the type of the value being bound, because it's not necessarily determined by the statement. But in the append case the type is always known, so specifying it again, via an explicit `duckdb_value`, is unnecessary.

You point out that appending can be done using `duckdb_append_data_chunk` instead of either type-specific append functions or (the proposed) `duckdb_append_value`. This is true, but it requires significantly more advanced and complex code from the caller. Additionally, making this work in higher-level languages can be very challenging. As an example, in Node, while it is possible to return an external buffer through the Node Add-on API, in [some environments](https://github.com/nodejs/node-addon-api/blob/main/doc/external_buffer.md) this buffer may be a copy, so that writing to it doesn't have the desired effect of altering the original memory, such as a data chunk's underlying vectors. As a result, although I've exposed appending data chunks through this API, it's not guaranteed to work end-to-end. To make it work, I'll likely need to implement some special C functions to write data to buffers, rather than doing this all from the JS side.

Even once that is addressed, doing the writes requires detailed understanding of the vector format for each type. This is already needed for reads; adding it for writes will significantly increase the necessary code and thus the complexity of each higher-level client. The type-specific append methods already encapsulate this logic, so it seems like less code to write and maintain to use them rather than reimplementing that logic in each client.

Regarding the functions that convert to and from structs: I implemented these mainly because the tests depend on them. It's also the case that tests for higher-level clients (such as the new Node client) are easier to write with these sorts of helpers. While it's true they could be implemented in the client without too much trouble, there are some subtleties with the conversion that are easy to get wrong (such as overflow). It seems that implementing this in each client creates duplication and additional chances to get it wrong or be inconsistent, while exposing them through the C API is straightforward and robust.

As a general point, I was surprised to hear that the append methods are not intended to be complete. Is this true for the bind methods as well? Based on the structure of the C API, I would have expected both of these sets to be complete. That would provide a consistent and dependable way to perform these tasks (appending & binding). If each type may or may not have a type-specific helper - if it may require using `duckdb_value` or direct writes to a data chunk instead - then that increases the friction and complexity of using the API, both for direct users of the C API and for implementors and users of higher-level language clients based on it.

I would prefer for these sets to be complete. But, if that's not to be, then it would be helpful to have a principle to guide when a type should or should not have a type-specific helper. Common sense says it would be based on usage - but I'm not confident I understand which types are commonly used. When I started this work, I assumed that the non-standard timestamp values are not commonly used. However, I've seen [multiple](https://discord.com/channels/909674491309850675/1148659944669851849/1308732109955006525) [threads](https://discord.com/channels/909674491309850675/921073327009853451/1308584628197785711) in the Discord recently about these types, so I'm now doubtful of that assumption.

If you'd still prefer for the sets of type-specific bind and append functions to not be complete, then I'll alter this PR. But I do think the C API would be better if they were complete, for the reasons I outline above. Please let me know your thoughts.

(Thanks for reviewing, and thanks reading this far. Sorry for the long reply; I've tried to express my thoughts as briefly and clearly as possible.)

**Comment by Mytherin** [comment]
These methods being complete is fundamentally not possible. The `_append_XX` and `_bind_XX` methods **cannot** be complete because DuckDB supports arbitrarily nested types. They also cannot support types defined in extensions. At best you could aim to support "all non-nested non-extension types", but that seems rather arbitrary.

While supporting basic types such as integers is fine in these methods - I'm not a big fan of extending them to support some arbitrary subset of types here. I'd prefer to keep the C API shorter and more elegant if the same functionality can be achieved using existing methods, and having methods such as `duckdb_append_value` or `duckdb_bind_value` as a fallback for non-supported types.

My view point on this is if you are intending to support all types then you need to use `duckdb_value` or `duckdb_data_chunk` anyway - since there is no way around this for arbitrarily nested types.

* `duckdb_value` then exists to support use cases where performance does not matter (e.g. binding individual values to a statement)
* `duckdb_data_chunk` exists to support performance sensitive use cases 

While I can see that `duckdb_data_chunk` might be complex to get working from Node, it is really the only way of achieving high-performance result input and output in DuckDB through the C API, so I don't really see a way around making that work in any case if high performance has to be achieved. The `duckdb_append_XXX` methods will be many times slower, with or without constructing a `duckdb_value`.

**Comment by jraymakers** [comment]
Thanks for the context. I would have suggested covering all non-nested types, but I didn't consider extension types; that's a good point.

It's also a good point that support writing to data chunks is the only way to achieve the best performance, and that the performance costs of `duckdb_value` likely aren't significant in most cases where they'll get used, assuming that data chunk writing is available when needed.

I'd still prefer to have access to these helpers, from an API consumer point-of-view, but I can understand wanting to keep the C API lean, from a maintainer point-of-view.

I'll update this PR, and also focus my future efforts on using `duckdb_value` and supporting writing to data chunks.

Thanks again for the quick response!

**Comment by jraymakers** [comment]
Posted a new [PR](https://github.com/duckdb/duckdb/pull/14954) with just the value creation/access and finiteness check functions. There were a few fixes needed to get these to work (see PR description). @Mytherin @taniabogatsch

**Comment by jraymakers** [comment]
Closed in favor of https://github.com/duckdb/duckdb/pull/14954

**Pull request closed without merging**
